File: android/app/src/main/java/com/prince/turtlekeyboard/ai/AlphaMatte.java
package com.prince.turtlekeyboard.ai;

import android.graphics.Bitmap;

/**
 * Difference-matting math: recovers per-pixel alpha from two model renders of the
 * same subject on pure white and pure black backgrounds. Shared by GifIntegration
 * and StickerIntegration; each caller supplies its own pass-2 prompt.
 *
 * <p>Algorithm: {@code alpha = 1 - euclidean(onWhite, onBlack) / sqrt(3·255²)};
 * true RGB back-solved from the black render as {@code observed / alpha}.
 */
public final class AlphaMatte {

    /** Snap thresholds clean up sub-pixel model-side RGB drift at the two ends. */
    private static final double SNAP_OPAQUE = 0.95;
    private static final double SNAP_TRANS  = 0.05;

    /** Max corner-pixel RGB drift from the expected background before a pass is
     *  treated as invalid; beyond this the matte collapses to garbage. */
    public static final int MAX_BG_DRIFT = 105;

    private AlphaMatte() {}

    /**
     * Max euclidean RGB distance between the bitmap's four corners and the
     * expected background. Callers use this to detect a pass where the model
     * ignored the background instruction and bail to a non-matte fallback.
     */
    public static int maxCornerDistance(Bitmap bitmap, int er, int eg, int eb) {
        int w = bitmap.getWidth(), h = bitmap.getHeight();
        int[] corners = {
                bitmap.getPixel(0, 0),
                bitmap.getPixel(w - 1, 0),
                bitmap.getPixel(0, h - 1),
                bitmap.getPixel(w - 1, h - 1),
        };
        int max = 0;
        for (int c : corners) {
            int dr = ((c >> 16) & 0xff) - er;
            int dg = ((c >> 8)  & 0xff) - eg;
            int db = ( c        & 0xff) - eb;
            int d = (int) Math.sqrt(dr * dr + dg * dg + db * db);
            if (d > max) max = d;
        }
        return max;
    }

    /**
     * Recovers per-pixel alpha and subject RGB from white/black-background renders.
     * Caller must ensure matching dimensions; passed-in bitmaps are not recycled.
     *
     * @param onWhite render with the locked white background
     * @param onBlack render with the locked black background
     * @return a new {@code ARGB_8888} bitmap with recovered alpha
     */
    public static Bitmap differenceMatte(Bitmap onWhite, Bitmap onBlack) {
        int w = onWhite.getWidth();
        int h = onWhite.getHeight();
        final double BG_DIST = Math.sqrt(3.0 * 255.0 * 255.0);

        int[] whitePix = new int[w * h];
        int[] blackPix = new int[w * h];
        onWhite.getPixels(whitePix, 0, w, 0, 0, w, h);
        onBlack.getPixels(blackPix, 0, w, 0, 0, w, h);

        int[] out = new int[w * h];
        for (int i = 0; i < whitePix.length; i++) {
            int pw = whitePix[i];
            int pb = blackPix[i];
            int rw = (pw >> 16) & 0xff, gw = (pw >> 8) & 0xff, bw = pw & 0xff;
            int rb = (pb >> 16) & 0xff, gb = (pb >> 8) & 0xff, bb = pb & 0xff;
            int dr = rw - rb, dg = gw - gb, db = bw - bb;
            double pixDist = Math.sqrt(dr * dr + dg * dg + db * db);
            double alpha = 1.0 - (pixDist / BG_DIST);
            if      (alpha > SNAP_OPAQUE) alpha = 1.0;
            else if (alpha < SNAP_TRANS)  alpha = 0.0;
            else if (alpha < 0.0)         alpha = 0.0;
            else if (alpha > 1.0)         alpha = 1.0;

            int a = (int) Math.round(alpha * 255.0);
            int r, g, b;
            if (alpha <= 0.0) {
                // Zero color on fully-transparent pixels so the encoder doesn't waste a palette slot.
                r = 0; g = 0; b = 0;
            } else {
                // observed = α·subject ⇒ subject = observed/α.
                r = (int) Math.min(255.0, Math.round(rb / alpha));
                g = (int) Math.min(255.0, Math.round(gb / alpha));
                b = (int) Math.min(255.0, Math.round(bb / alpha));
            }
            out[i] = (a << 24) | (r << 16) | (g << 8) | b;
        }

        Bitmap result = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
        result.setPixels(out, 0, w, 0, 0, w, h);
        return result;
    }
}

File: ios/TurtleKeyboardExtension/AI/AlphaMatte.swift
import Foundation
import CoreGraphics
import ImageIO

/// Difference-matting math shared between `/sticker` and `/gif`. Given two
/// renders of the same scene — one on pure white, one on pure black —
/// recovers per-pixel alpha and the subject's true RGB.
///
/// Direct port of `android/.../ai/AlphaMatte.java`. Same constants, same
/// algorithm, same snap thresholds.
///
/// ```
///   pixelDist = euclidean(pixOnWhite, pixOnBlack)
///   bgDist    = euclidean(WHITE, BLACK) = sqrt(3 · 255²) ≈ 441.67
///   alpha     = 1 - (pixelDist / bgDist)
/// ```
/// Subject pixels match across renders ⇒ `alpha = 1`. Background pixels
/// differ by the full distance ⇒ `alpha = 0`. Anti-aliased edges land at
/// intermediate values. True RGB is back-solved from the black render:
/// `observed = α · true` ⇒ `true = observed / α`. Snap thresholds clean
/// up small model-side RGB drift at the two ends.
///
/// Article reference:
/// `jidefr.medium.com/generating-transparent-background-images-with-nano-banana-pro`
enum AlphaMatte {

    /// Snap-to-opaque / snap-to-transparent thresholds on recovered alpha.
    /// Tiny model-side RGB drift on subject pixels otherwise leaves them
    /// at ~0.97 alpha (faint translucency); pure-background pixels
    /// otherwise sit at ~0.02 (faint ghost). Snapping cleans both ends.
    private static let snapOpaque: Double = 0.95
    private static let snapTrans:  Double = 0.05

    /// Background euclidean distance — `sqrt(3·255²)`.
    private static let bgDist: Double = (3.0 * 255.0 * 255.0).squareRoot()

    /// Compute the matte. Returns nil when dimensions differ, the buffers
    /// can't be drawn, or the output `CGImage` can't be constructed.
    /// Caller owns no input/output lifetime beyond the returned `CGImage`.
    static func differenceMatte(onWhite: CGImage, onBlack: CGImage) -> CGImage? {
        let w = onWhite.width
        let h = onBlack.height
        guard w > 0, h > 0,
              w == onBlack.width, h == onWhite.height else {
            return nil
        }

        // Decode both inputs into known-layout RGBA8 buffers. We use
        // `premultipliedLast` for the *input* contexts because CGContext
        // requires a premultiplied alpha format — but since the inputs are
        // opaque PNGs (alpha == 255 everywhere), premultiplied == raw RGB
        // so the channel values are correct.
        guard let whiteBuf = readRGBA8(from: onWhite, width: w, height: h),
              let blackBuf = readRGBA8(from: onBlack, width: w, height: h) else {
            return nil
        }

        let count = w * h * 4
        var out = [UInt8](repeating: 0, count: count)

        // Walk pixels. Hot loop — keep math in Double; the inner branches
        // are predictable and the compiler vectorises the channel reads.
        whiteBuf.withUnsafeBufferPointer { wp in
            blackBuf.withUnsafeBufferPointer { bp in
                out.withUnsafeMutableBufferPointer { op in
                    var i = 0
                    while i < count {
                        let rw = Double(wp[i]),     gw = Double(wp[i + 1]), bw = Double(wp[i + 2])
                        let rb = Double(bp[i]),     gb = Double(bp[i + 1]), bb = Double(bp[i + 2])
                        let dr = rw - rb, dg = gw - gb, db = bw - bb
                        let pixDist = (dr * dr + dg * dg + db * db).squareRoot()
                        var alpha = 1.0 - (pixDist / bgDist)
                        if      alpha > snapOpaque { alpha = 1.0 }
                        else if alpha < snapTrans  { alpha = 0.0 }
                        else if alpha < 0.0        { alpha = 0.0 }
                        else if alpha > 1.0        { alpha = 1.0 }

                        if alpha <= 0.0 {
                            // Fully transparent — zero RGB so downstream
                            // PNG / GIF encoders don't waste palette slots
                            // on never-visible colors.
                            op[i] = 0; op[i + 1] = 0; op[i + 2] = 0; op[i + 3] = 0
                        } else {
                            // From the black render: observed = α · subject
                            //   ⇒ subject = observed / α
                            // Stored UN-premultiplied (matches Android's
                            // ARGB_8888 contract and the output CGImage's
                            // `.last` alpha layout below).
                            let r = min(255.0, (rb / alpha).rounded())
                            let g = min(255.0, (gb / alpha).rounded())
                            let b = min(255.0, (bb / alpha).rounded())
                            op[i]     = UInt8(r)
                            op[i + 1] = UInt8(g)
                            op[i + 2] = UInt8(b)
                            op[i + 3] = UInt8((alpha * 255.0).rounded())
                        }
                        i += 4
                    }
                }
            }
        }

        // Wrap the output bytes in an un-premultiplied RGBA8 CGImage. We
        // use `.last` (no premultiplication) here — CGImage initialisers
        // accept it, even though CGContext drawing wouldn't. Downstream
        // PNG / GIF encoders read the buffer as-is.
        let bytesPerRow = w * 4
        guard let provider = CGDataProvider(data: Data(out) as CFData) else {
            return nil
        }
        let bitmapInfo = CGBitmapInfo(rawValue:
            CGImageAlphaInfo.last.rawValue | CGBitmapInfo.byteOrder32Big.rawValue)
        return CGImage(
            width: w,
            height: h,
            bitsPerComponent: 8,
            bitsPerPixel: 32,
            bytesPerRow: bytesPerRow,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: bitmapInfo,
            provider: provider,
            decode: nil,
            shouldInterpolate: false,
            intent: .defaultIntent
        )
    }

    /// Decode a `CGImage` into a fresh RGBA8 buffer of size `width*height*4`.
    /// Uses a `premultipliedLast` CGContext because that's what CGContext
    /// requires; for opaque inputs the result is the same as un-premul.
    private static func readRGBA8(from image: CGImage, width w: Int, height h: Int) -> [UInt8]? {
        var buf = [UInt8](repeating: 0, count: w * h * 4)
        let bytesPerRow = w * 4
        let bitmapInfo: UInt32 = CGImageAlphaInfo.premultipliedLast.rawValue
            | CGBitmapInfo.byteOrder32Big.rawValue
        let cs = CGColorSpaceCreateDeviceRGB()
        let result: CGContext? = buf.withUnsafeMutableBytes { raw -> CGContext? in
            guard let base = raw.baseAddress else { return nil }
            return CGContext(
                data: base,
                width: w,
                height: h,
                bitsPerComponent: 8,
                bytesPerRow: bytesPerRow,
                space: cs,
                bitmapInfo: bitmapInfo
            )
        }
        guard let ctx = result else { return nil }
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: w, height: h))
        return buf
    }
}
